### ISSUE TITLE
[Organisation] _Early_ feature freeze for git master prior to release + counter-proposal(s)

### ISSUE THREAD

**Issue by keyspace** [CleverRaven/Cataclysm-DDA#20338]
**TL;DR:** Stop using `git master` to stage all the feature branches at once. Deem some "worthy for next release" during pre-release phase, then merge after release.

Current situation: not-yet-merged feature branches are use-tested as much as bugfixes for already-merged feature branches. That is, at best by the people writing/merging them, and maybe someone interested in the feature who can build manually. Anyway, a feature branch is not likely to get enough people use-testing it - at least not enough to find regressions fast.

However, C:DDA currently has an unusual number of people using experimental auto-builds. Don't have precise numbers, but [could be 1000+](https://www.reddit.com/r/cataclysmdda/comments/5gnpwb/players_survey_2016_results/). (Should expect that number to drop after a release.)

Say, if a feature branch was deemed "good enough for general testing", it could be "slated for post-release", e.g. its milestone set to `0.E`, and soon after `0.D` it could be merged to `git master`.

Not sure if still applicable, but [feature freeze prior to release](http://smf.cataclysmdda.com/index.php?topic=4104.0) would then be a good time to review feature branches.

### Pros

* Allows having more frequent releases (but see cons).
* Having that could allow mod authors to advertise compatibility with (hopefully) something recent, and to themselves have a less hap-hazard development cycle.
* Potentially allows moving some mods "out" of the C:DDA repo. E.g. as a git submodule referencing a separate repo with "mod manager" describing mods in JSON, a-la [CKAN](https://github.com/keyspace/CKAN#whats-the-ckan). (Haven't though on this much yet.)

### Cons

* Slows down C:DDA development by introducing an org-level delay, especially for feature inclusion, and possibly refactoring.
* (added from discussion) Old PRs take more work to merge, and are at risk of becoing unmergeable, with further slows development and discourages contributors.
* Very likely to reduce the number of people playing experimental builds, and by extension testing, therefore further slowing down development.
* Requires planning, i.e. paperwork, and enforcement, i.e. more paperwork.

-----

Started when [first describing](https://github.com/CleverRaven/Cataclysm-DDA/issues/20251#issuecomment-279025706) the #20251 issue ("[Proposal] Revert last month's input context handling changes"), picked up after [describing the stall with bionic slots](https://github.com/CleverRaven/Cataclysm-DDA/issues/20305#issuecomment-281327417) in issue #20305 ("Hide or implement bionic slot restrictions for 0.D").

**Comment by kevingranade** [comment]
Title's a little misleading, seems like it should be replaced by some version of your tl;dr.

In other words, you're proposing an early feature freeze, there's no continuous process, just refuse to merge some PRs until after a release.

It would accelerate progress toward the next release, at least at first.  Unfortunately letting PRs build up tends to be rather discouraging for contributors, and the workload after you open the floodgates again is crushing.
Neither of the other pros is actually related to the proposed feature freeze, but rather to just getting a release out.

There are some additional cons as well:
Old PRs take more work to merge, and are at risk of becoing unmergeable, with further slows development and discourages contributors.

Counter-proposal:
# Release branch
We cut a branch for 0.D earlier than usual, before it's fully stable, before we would do a feature freeze.  Then merge just bugfixes into the release branch, and continue merging both bugfixes and features into mainline.

Pros:
* No halt to feature development.
* Same benefits of earlier release.

Cons:
* More coordination overhead, similar to an early feature freeze.
* Strictly duplicates the work of merging bugfixes, though merging the same branch twice is much easier than merging two seperate branches.
* AFAIK github has no features for supporting this, people merging bugfixes will have to make totally sure they merge to both development branches, because once one merge is pushed, I think it will close the related PR.
* Some players will shift to the release once it happens, that's a given, but I don't think a critical mass of players will, especially if the experimental keeps having features added even while the release branch is being stabilized.

**Comment by Coolthulhu** [comment]
> We cut a branch for 0.D earlier than usual, before it's fully stable, before we would do a feature freeze.

How about releasing two 0.Ds? One would be pre-release, released as soon as we have no milestones, the other one would be the proper one with fixes.
As the first one would be feature-frozen, there would be no risk of players losing out stuff or switching to a less balanced version. Mods aiming for 0.D would most likely work with either version.

**Comment by kevingranade** [comment]
I usually do a point release if there are any serious issues found after the first one, it's really likely this time as there will likely be qute a surge of people trying out the new stable when we release it.
In other words, yea we can do a series of releases for 0.D.whatever, though I'd rather not release with any known major issues.

**Comment by keyspace** [comment]
Edited issue title.

-----

> Unfortunately letting PRs build up tends to be rather discouraging for contributors, and the workload after you open the floodgates again is crushing.

Agreed. Wanted to get this off my drive - easier to disprove if not doing it alone.

-----

"Two `0.D`s" scheme would still require duplication for bugfix merges, yes? Maybe it's not so bad - hard to tell IMO, depends on the number of bugs that get reported for pre-release.

**Comment by keyspace** [comment]
OP seems to be unsupported - think I understand why, and tend to agree.

Ultimately, the choice to have a release branch, or make a point release, is up to project owners/collaborators.

Closing this zombie proposal - ping if there's anything left to discuss, will reopen.
